### DISCUSSION TITLE
Make @WithSpan apply to Class level.

### DISCUSSION THREAD

**Discussion by bmistry13** [open-telemetry/opentelemetry-java-instrumentation#2164] [Q&A]
@WithSpan annotation is only applied to METHOD.  Can we change behavior to apply to Class so ALL methods will get instrumented it is very annoying to add for each method?

**Comment by iNikem** [comment]
Do you really mean ALL methods or all public methods? What about inherited methods?

**Reply by bmistry13** [reply]
I mean all public methods from the current class.  If you want inherited methods as well @WithSpan(inherited=true) or something similar.

[Follow similar to Spring Transactional](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Transactional.html)

**Comment by bmistry13** [comment]
@iNikem ,  I tried the @WithSpan method level but agent 0.14 does not pick a method at all.  I am in spring boot application and I have Interface A and Concrete Implementation class but Agent does not seem to pick @WithSpan and does not create span.  Is there anything I need to do besides annotation?

**Reply by trask** [reply]
hi @bmistry13, did you put the annotation on the interface method or concrete method? If you put it on the interface method previously, does it work if you move it to the concrete method?

**Reply by bmistry13** [reply]
@trask , I put it on Concrete  Implementation but Spring beans may do a Dynamic PROXY pattern and still NOT showing.  Is there anything I need to do? Please let me know.

**Reply by trask** [reply]
hey @bmistry13 can you create a small repro and open an issue for this?

**Reply by bmistry13** [reply]
I will try to reproduce it in a simple standalone spring-boot app.

**Comment by sp710419** [comment]
I have 2 springboot applications instrumented with otel javaagent library. with auto instrumentation everything works fine. Springboot A calls SB B. and B intern makes some internal DB calls. All spans and traces I could see in my jaeger backend. Now I want to inject a custom attribute to the existing trace. 

```java
@RestController
public class AppAController {
    
    @Autowired
    private RestTemplate restTemplate;

    @RequestMapping("/test-tracing")
    @WithSpan
    public String entryPointController()  {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
        HttpEntity<String> entity = new HttpEntity<>(headers);
       
        // Here I am calling SB2, second mS endpoint 
        String response=restTemplate.exchange("http://opentelemetryappb:9082/service", HttpMethod.GET, entity, String.class).getBody();

        return "Response from opentelemetry-app-b => : "+response;
    }
```

As per the above discussion, I just added @WIthSpan to this controller mehod. I understand I have to write some code to create a custom span here using OTEL library. Can some one pls help here how do I inject custom attributes to the existing spans here ? Any reference project/code ??

My springboot using embedded tomcat..

**Reply by mateuszrzeszutek** [reply]
Hey @sp710419 ,
Do you want to just set custom attributes to an already existing span? Or add a completely new span to the trace?
If it's the first case, you don't need to use `@WithSpan` at all - you can set attributes using `Span.current()`. For example:
```java
    @RequestMapping("/test-tracing")
    public String entryPointController()  {
        Span.current().setAttribute("my-attribute", 42);

        // ...
    }
```
You should use `@WithSpan` only when you want to create a new span for some operation.

**Reply by sp710419** [reply]
Hi @mateuszrzeszutek  thank you for your response..
Well this was the first thing which I tried. Without annotation, just added a span attribute. But it didn’t work. The idea was to add a custom attributed at the entrypoint so that the attribute should be available now on throughout the trace across microservices. When my mSA -> mSB -> mSC, this attribute should be inherited to all the calles from start to end ..As I’m exporting the data to multiple datasources like jaeger, MS app insight etc..I can filter that s attribute which is specific to my application only. As I said earlier, my application is instrumented with otel all-in-obe agent which does all work of generating spans and traces.
If you have any similar project for reference, it will be great .. TIA

**Reply by mateuszrzeszutek** [reply]
> When my mSA -> mSB -> mSC, this attribute should be inherited to all the calles from start to end

Ah, that looks like a use case for `Baggage` -- please take a look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/5070#discussioncomment-1955419, I've answered a similar problem there. Does that help you?

**Reply by sp710419** [reply]
Thank you. Yes, earlier I tried following your reference project and used a DemoSpanProcessor .
here was my sample code.
..
@AutoService(DemoSpanProcessor.class)
public class DemoSpanProcessor implements SpanProcessor {

    @WithSpan
    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        Baggage.fromContext(parentContext).forEach((s, baggageEntry) -> span.setAttribute(s, baggageEntry.getValue()));
    }
....

@AutoService(SdkTracerProviderConfigurer.class)
public class DemoSdkTracerProviderConfigurer implements SdkTracerProviderConfigurer {
    @Override
    public void configure(SdkTracerProviderBuilder tracerProvider, ConfigProperties config) {
        tracerProvider
                .addSpanProcessor(new DemoSpanProcessor());
    }
}
.......

Then I wrote a interceptor for my entrypoint mS and injected the custom attribute..

@Component
public class ApplicationInterceptor implements HandlerInterceptor {

    @Override
    public void postHandle(
            HttpServletRequest request,
            HttpServletResponse response,
            Object handler,
            ModelAndView modelAndView) {

            OpenTelemetry otel = GlobalOpenTelemetry.get();
            Tracer tracer = otel.getTracerProvider().get("opentelemetry-aap-a");
            Baggage.current().toBuilder().put("transactionId","123456");
    }
}
...
..

This approach also did not work. 
Sorry I am new to OTEL..Not sure If I am doing something wrong..


TIA..

**Reply by mateuszrzeszutek** [reply]
> `Baggage.current().toBuilder().put("transactionId","123456");`

`Baggage` is immutable - you have to start (and close) a new scope every time you want to change that. For example:

```java
Baggage newBaggage = Baggage.current().toBuilder().put("transactionId","123456");
try (Scope scope = newBaggage.makeCurrent()) {
  // ... code that calls downstream services
}
```

**Reply by sp710419** [reply]
Thank you @mateuszrzeszutek , Appricate your time and responses..

Tried what you suggested and still did not work.

    @RequestMapping("/test-tracing")
    public String entryPointController() {
         ...
        String response = null;
        Baggage newBaggage = Baggage.current().toBuilder().put("transId","12345678").build();
        try (Scope scope = newBaggage.makeCurrent()) {
            
            //calling another mS
            response=restTemplate.exchange("http://opentelemetryappb:9082/service", HttpMethod.GET, entity, String.class).getBody();
        }
        return "Response from opentelemetry-app-b => : "+response;
    }

I still do not see this custom attribute in my traces..
Do I need to add any other jar along with javaagent ? Right now, I have only auto agent added to my classpath in Dockerfile

-Dapplication.name=opentelemetry-app-a -javaagent:/opentelemetry-javaagent-all.jar -jar

TIA again !!

**Reply by mateuszrzeszutek** [reply]
Hmm, did you put the custom `SpanProcessor` directly into your application code? Unfortunately that's currently not supported by the agent - you have to build a separate [extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension) jar that'll contain the `SpanProcessor` and `SdkTracerProviderConfigurer` implementations. Please take a look at the extension page - you can just copy the project and modify it so that it suits your needs.

**Reply by sp710419** [reply]
Thanks @mateuszrzeszutek !! It worked.

**Comment by MrNocTV** [comment]
Hi guys, is this feature implemented? 
I want to collect metrics info of my applications. However the default metrics is not enough. I want to apply @WithSpan on all methods and classes in my application. Is this possible?

**Reply by mateuszrzeszutek** [reply]
Hey @MrNocTV ,
No, it hasn't been implemented yet. @iNikem's [comment](https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/2164#discussioncomment-329961) mentions one problem/caveat with that, but there's a lot more really: should we instrument public methods? what about protected, private ones? what about inherited methods? private methods in the parent class? (also see #2574, since it's somewhat similar)
This feature is quite complex, and there's no "natural" answer to the questions above, it varies case by case.
